### PR TITLE
Added end-to-end unit tests for the attested TLS feature

### PR DESCRIPTION
**Pull request by soccerGB**
Unittests Transport Layer Security end-2-end scenarios for https://github.com/microsoft/openenclave/issues/1325
=====================

Test scenarios:

In this test, we have three test hooks for enabling test scenarios.

- fail_cert_verify_callback
   When set to true, it will force a tls connecting party to fail cert_verify_callback unconditionally.
   When set to false, cert_verify_callback will perform normally.

- fail_enclave_identity_verifier_callback;
   When set to true, it will force a tls connecting party to fail fail_enclave_identity_verifier_callback unconditionally.
   When set to false, enclave_identity_verifier_callback will perform normally.

- fail_oe_verify_attestation_certificate;
   When set to true, it simulates a failure return from calling oe_verify_attestation_certificate
   When set to false, fail_oe_verify_attestation_certificate will perform normally.

Here the scenarios covered in these tests via running run_scenarios_tests()

1. A positive scenario
    This is the case we hope always happen when both TLS connecting parties successfully validate each other's certificate and enclave identity.
    And everything works correctly all test hooks mentioned above set to false.

2. Negative scenarios

    We run tls connection tests three times, each with a tls connecting party set to true one of the following hook configurations while the other connecting party with all hooks set to false.

    fail_cert_verify_callback
    fail_enclave_identity_verifier_callback
    fail_oe_verify_attestation_certificate

    And we ran above three scenarios for both server and client roles. That got us six negative test scenarios.

### CODE REVIEW

File: tests/tls_e2e/common/utility.cpp [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        result,
+        "oe_get_private_key_by_policy(OE_SEAL_POLICY_UNIQUE) = %s",
+        oe_result_str(result));
+
+done:
+    return result;
+}
+
+// Compute the sha256 hash of given data.
+static int Sha256(const uint8_t* data, size_t data_size, uint8_t sha256[32])
```
Comment (radhikaj) [inline tests/tls_e2e/common/utility.cpp]: Should this be a lower case sha256

File: tests/tls_e2e/common/utility.cpp [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    }
+    ret = true;
+done:
+    if (signer)
+        free(signer);
+
+    if (modulus != NULL)
+        free(modulus);
+
+    mbedtls_pk_free(&ctx);
```
Comment (radhikaj) [inline tests/tls_e2e/common/utility.cpp]: If signer is NULL, you will go to done without initing ctx. And then you are freeing it here before it was inited.

File: tests/tls_e2e/common/utility.cpp [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        modulus[i] = modulus[modulus_size - 1 - i];
+        modulus[modulus_size - 1 - i] = tmp;
+    }
+
+    // Calculate the MRSIGNER value which is the SHA256 hash of the
+    // little endian representation of the public key modulus. This value
+    // is populated by the signer_id sub-field of a parsed oe_report_t's
+    // identity field.
+
+    OE_TRACE_ERROR("modulus_size=%ld\n", modulus_size);
```
Comment (radhikaj) [inline tests/tls_e2e/common/utility.cpp]: Maybe this should be TRACE_INFO

File: tests/tls_e2e/common/utility.cpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+    result = oe_verify_attestation_certificate(
+        cert_buf, cert_size, enclave_identity_verifier, NULL);
+    OE_CHECK_MSG(
+        result,
+        "oe_verify_attestation_certificate failed with result = %s\n",
+        oe_result_str(result));
+
+    OE_TRACE_INFO("\nReturned from oe_verify_attestation_certificate\n");
+    ret = 0;
```
Comment (radhikaj) [inline tests/tls_e2e/common/utility.cpp]: Maybe ret shoud be initialized to a failure value instead of 0?

File: tests/tls_e2e/host/host.cpp [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        !server_config->args.fail_oe_verify_attestation_certificate)
+    {
+        OE_TEST(g_client_thread_exit_code == 0);
+    }
+    else
+    {
+        // g_client_thread_exit_code could be any values in negative test cases
+        g_client_thread_exit_code = 0;
+    }
+    pthread_exit((void*)&g_client_thread_exit_code);
```
Comment (radhikaj) [inline tests/tls_e2e/host/host.cpp]: Would you want to do this after done so it is hit in all paths?

File: tests/tls_e2e/common/utility.cpp (hunk 2 above)
Comment (soccerGB) [reply tests/tls_e2e/common/utility.cpp]: Good catch. Addressed by moving the ctx initialization up

File: tests/tls_e2e/common/utility.cpp (hunk 1 above)
Comment (soccerGB) [reply tests/tls_e2e/common/utility.cpp]: fixed.

File: tests/tls_e2e/common/utility.cpp (hunk 3 above)
Comment (soccerGB) [reply tests/tls_e2e/common/utility.cpp]: fixed

File: tests/tls_e2e/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+set(TEST_NAME "tests/tls_e2e_host")
+set(HOST_FILE "tls_e2e_host")
+set(ENC_FILE "tls_server_enc")
+set(ENC_FILE2 "tls_client_enc")
+
+add_enclave_test(tests/tls_e2e
+	tls_e2e_host tls_server_enc $<TARGET_FILE:tls_client_enc>)
+
+set_tests_properties(tests/tls_e2e PROPERTIES SKIP_RETURN_CODE 2)
```
Comment (anitagov) [inline tests/tls_e2e/CMakeLists.txt]: Missing newline at EOF

File: tests/tls_e2e/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+Unittests Transport Layer Security end-2-end scenarios
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: add space between Unit and tests

File: tests/tls_e2e/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+=====================
+
+Test scenarios:
+
+In this test, we have three test hooks for enabling test scenarios.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: How about rephrasing to: 
Three test hooks are available for enabling test scenarios.

File: tests/tls_e2e/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+- fail_cert_verify_callback
+   When set to true, it will force a tls connecting party to fail cert_verify_callback unconditionally.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: For consistency, make tls as TLS in the document

File: tests/tls_e2e/README.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+Unittests Transport Layer Security end-2-end scenarios
+=====================
+
+Test scenarios:
+
+In this test, we have three test hooks for enabling test scenarios.
+
+- fail_cert_verify_callback
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: Perhaps, numbering them as 1), 2) and 3) makes sense

File: tests/tls_e2e/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+- fail_enclave_identity_verifier_callback;
+   When set to true, it will force a tls connecting party to fail fail_enclave_identity_verifier_callback unconditionally.
+   When set to false, enclave_identity_verifier_callback will perform normally.
+
+- fail_oe_verify_attestation_certificate;
+   When set to true, it simulates a failure return from calling oe_verify_attestation_certificate
+   When set to false, fail_oe_verify_attestation_certificate will perform normally.
+
+Here the scenarios covered in these tests via running run_scenarios_tests()
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: missing "are" after Here

File: tests/tls_e2e/client_enc/client.cpp [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/tests.h>
+
+#include <openenclave/corelibc/arpa/inet.h>
+#include <openenclave/corelibc/netdb.h>
+#include <openenclave/corelibc/netinet/in.h>
+#include <openenclave/internal/posix/device.h>
+
+//#include <bits/stdfile.h> // For stderr & FILE
+#include <errno.h> // For errno & error defs
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Cleanup the commented out stuff if you don't need them

File: tests/tls_e2e/client_enc/client.cpp [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+    int launch_tls_client(
+        struct tls_control_args* config,
+        char* server_name,
+        char* server_port);
+    int setup_tls_server(struct tls_control_args* config, char* server_port);
+};
+
+struct tls_control_args g_control_config;
+
+// This is the identity validation callback. An TLS connecting party (client or
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Change "An TLS" to "A TLS"

File: tests/tls_e2e/client_enc/client.cpp [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+// server) can verify the passed in "identity" information to decide whether to
+// accept an connection reqest
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: typo "reqest" - request
Could add: "or not" at the end

File: tests/tls_e2e/client_enc/client.cpp [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+    if (identity->security_version < 1)
+    {
+        OE_TRACE_ERROR(
+            "identity->security_version checking failed (%d)\n",
+            identity->security_version);
+        goto done;
+    }
+
+    // Dump an enclave's unique ID, signer ID and Product ID. They are
+    // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves In a real scenario,
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Add a "full-stop" after SGX enclaves

File: tests/tls_e2e/README.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+1. A positive scenario
+    This is the case we hope always happen when both TLS connecting parties successfully validate each other's certificate and enclave identity.
+    And everything works correctly all test hooks mentioned above set to false.
+
+2. Negative scenarios
+
+    We run tls connection tests three times, each with a tls connecting party set to true one of the following hook configurations while the other connecting party with all hooks set to false.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: Change instances of "tls" to TLS

File: tests/tls_e2e/client_enc/client.cpp [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+    // custom id checking should be done here
+    OE_TRACE_INFO("\nidentity->signer_id :\n");
+    for (int i = 0; i < OE_UNIQUE_ID_SIZE; i++)
+        OE_TRACE_INFO("0x%0x ", (uint8_t)identity->signer_id[i]);
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: This is incorrect: You should print the unique_id i.e. MRENCLAVE
identity->signer_id[i] should be identity->unique_id[i]

File: tests/tls_e2e/client_enc/client.cpp [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+        OE_TRACE_ERROR(
+            "identity->security_version checking failed (%d)\n",
+            identity->security_version);
+        goto done;
+    }
+
+    // Dump an enclave's unique ID, signer ID and Product ID. They are
+    // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves In a real scenario,
+    // custom id checking should be done here
+    OE_TRACE_INFO("\nidentity->signer_id :\n");
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Shouldn't this be identity->unique_id?

File: tests/tls_e2e/client_enc/client.cpp [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+    // certificate
+    //            can be provisioned. The server's preferences in its
+    //            CertficateRequest message will be ignored and our only cert
+    //            will be sent regardless of whether it matches those
+    //            preferences - the server can then decide what it wants to do
+    //            with it.
+    //
+    // The provided pk_key needs to match the public key in the first
+    // certificate in own_cert, or all handshakes using that certificate will
+    // fail. It is your responsibility to ensure that; this function will not
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Remove the ; after that

File: tests/tls_e2e/client_enc/client.cpp [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+    //            multiple certificates are adequate, preference is given to the
+    //            one set by the first call to this function, then second, etc.
+    // On client, only the first call has any effect. That is, only one client
+    // certificate
+    //            can be provisioned. The server's preferences in its
+    //            CertficateRequest message will be ignored and our only cert
+    //            will be sent regardless of whether it matches those
+    //            preferences - the server can then decide what it wants to do
+    //            with it.
+    //
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Formatting seems OFF - Is this intentional?

File: tests/tls_e2e/client_enc/client.cpp [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+    // The provided pk_key needs to match the public key in the first
+    // certificate in own_cert, or all handshakes using that certificate will
+    // fail. It is your responsibility to ensure that; this function will not
+    // perform any check. You may use mbedtls_pk_check_pair() in order to
+    // perform this check yourself, but be aware that this function can be
+    // computationally expensive on some key types.
+
```
Comment (anitagov) [inline tests/tls_e2e/client_enc/client.cpp]: Definitely nice explanation.

File: tests/tls_e2e/host/host.cpp [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+            "identity.security_version check failed (%d)\n",
+            identity->security_version);
+        goto done;
+    }
+
+    // the unique ID for the enclave
+    // For SGX enclaves, this is the MRENCLAVE value
+    OE_TRACE_INFO("identity->signer_id :\n");
+    for (int i = 0; i < OE_UNIQUE_ID_SIZE; i++)
+        OE_TRACE_INFO("0x%0x ", (uint8_t)identity->signer_id[i]);
```
Comment (anitagov) [inline tests/tls_e2e/host/host.cpp]: Error: Replace signer_id to unique_id in lines 100 and 102

File: tests/tls_e2e/common/utility.cpp (hunk 4 above)
Comment (soccerGB) [reply tests/tls_e2e/common/utility.cpp]: Good catch. Appropriate error code was set to MBEDTLS_ERR_X509_CERT_VERIFY_FAILED.

File: tests/tls_e2e/host/host.cpp (hunk 22 above)
Comment (soccerGB) [reply tests/tls_e2e/host/host.cpp]: good catch! Fixed

File: tests/tls_e2e/client_enc/client.cpp (hunk 20 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: Fixed

File: tests/tls_e2e/client_enc/client.cpp (hunk 19 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: Fixed

File: tests/tls_e2e/client_enc/client.cpp (hunk 18 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: good catch! fixed

File: tests/tls_e2e/client_enc/client.cpp (hunk 17 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: fixed

File: tests/tls_e2e/README.md (hunk 16 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: good idea.
addressed

File: tests/tls_e2e/client_enc/client.cpp (hunk 15 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: addressed

File: tests/tls_e2e/client_enc/client.cpp (hunk 14 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: addressed

File: tests/tls_e2e/client_enc/client.cpp (hunk 13 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: addressed

File: tests/tls_e2e/client_enc/client.cpp (hunk 12 above)
Comment (soccerGB) [reply tests/tls_e2e/client_enc/client.cpp]: Removed

File: tests/tls_e2e/README.md (hunk 11 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: added

File: tests/tls_e2e/README.md (hunk 10 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: addressed

File: tests/tls_e2e/README.md (hunk 9 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: changed

File: tests/tls_e2e/README.md (hunk 7 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: added

File: tests/tls_e2e/README.md (hunk 8 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: Changed as suggested

File: tests/tls_e2e/CMakeLists.txt (hunk 6 above)
Comment (soccerGB) [reply tests/tls_e2e/CMakeLists.txt]: added

**Comment by soccerGB** [comment]
bors try

File: tests/tls_e2e/common/utility.cpp [hunk 23]
```diff
@@ -0,0 +1,318 @@
...
+        &public_key_buf,
+        &public_key_buf_size,
+        &private_key_buf,
+        &private_key_buf_size);
+    OE_CHECK_MSG(result, " failed with %s\n", oe_result_str(result));
+
+#if 0
+    OE_TRACE_INFO("private_key_buf_size:[%ld]\n", private_key_buf_size);
+#endif
+
```
Comment (anitagov) [inline tests/tls_e2e/common/utility.cpp]: Remove the section lines:111 to 113

File: tests/tls_e2e/README.md [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+   When set to false, enclave_identity_verifier_callback will perform normally.
+
+3. fail_oe_verify_attestation_certificate;
+   When set to true, it simulates a failure return from calling oe_verify_attestation_certificate
+   When set to false, fail_oe_verify_attestation_certificate will perform normally.
+
+Here are the scenarios covered in these tests via running run_scenarios_tests()
+
+1. A positive scenario
+    This is the case we hope always happen when both TLS connecting parties successfully validate each other's certificate and enclave identity.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: line 23 - add "will" after hope

File: tests/tls_e2e/README.md [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    And everything works correctly all test hooks mentioned above set to false.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: couple of missing words. Guess this is what you want: And everything works correctly **when** all the test hooks mentioned above **are** set to false.

File: tests/tls_e2e/README.md [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+
+2. Negative scenarios
+
+    We run TLS connection tests three times, each with a TLS connecting party set to true one of the following hook configurations while the other connecting party with all hooks set to false.
```
Comment (anitagov) [inline tests/tls_e2e/README.md]: missing word after true?

File: tests/tls_e2e/README.md (hunk 25 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: addressed

File: tests/tls_e2e/README.md (hunk 24 above)
Comment (soccerGB) [reply tests/tls_e2e/README.md]: addressed

File: tests/tls_e2e/common/utility.cpp (hunk 23 above)
Comment (soccerGB) [reply tests/tls_e2e/common/utility.cpp]: This line will come in handy when debugging some signing issue, i will keep it.

**Pull request merged**